packages/block-library/src/navigation-link/edit.js

```diff
@@ -105,6 +117,39 @@ function NavigationLinkEdit( {
 		}
 	}, [ url ] );
 
+	useEffect( () => {
+		function handleDragStart( e ) {
+			// Check the first time when the dragging starts
+			handleDragEnter( e );
+		}
+
+		// Set isDragEnter to false whenever the user cancel the drag event by either releasing the mouse or press Escape.
+		function handleDragEnd() {
+			setIsDragEnter( false );
+		}
+
+		function handleDragEnter( e ) {
+			// Check if the current target is inside the item element
+			if ( listItemRef.current.contains( e.target ) ) {
+				setIsDragEnter( true );
+			} else {
+				setIsDragEnter( false );
+			}
+		}
+
+		// Bind these events to the document to catch all drag events.
+		// Ideally, we can also use `event.relatedTarget`, but sadly that doesn't work in Safari.
+		document.addEventListener( 'dragstart', handleDragStart );
+		document.addEventListener( 'dragend', handleDragEnd );
+		document.addEventListener( 'dragenter', handleDragEnter );
+
+		return () => {
+			document.removeEventListener( 'dragstart', handleDragStart );
+			document.removeEventListener( 'dragend', handleDragEnd );
+			document.removeEventListener( 'dragenter', handleDragEnter );
+		};
```

Comment:
Could these events be bound to the list Item itself? taking the example above, if the logic were wrapped up into a hook it could be done like this:
```jsx
const elementRef = useRef();
const [ isDraggingWithin, onDragStart, onDragEnter, onDragEnd ] = useIsDraggingWithin( ref );

return ( 
  <SomeElement
    ref={ elementRef }
    onDragStart={ onDragStart }
    onDragEnter={ onDragEnter }
    onDragEnd={ onDragEnd }
    className={ classnames( { 'is-dragging-within': isDraggingWithin } ) }
  />
);
```

Not sure though if the extra verbosity is worth making it more declarative?

Reply:
Probably not, we still need the drag event to handle "dragging-out" of the list item. Perhaps we can do it by adding a `onDragLeave` event though, I'll see what I can do.

Reply:
Nah, I don't think it would work, not without making it more complicated at least. 🤔 
`dragleave` event doesn't give us the event target that is dragging **to**, but only what we are leaving **from**. Hence, we cannot properly set `isDraggingWithin` to `false` when the user drag it outside the list item.